utilities/secondary_index/faiss_ivf_index.cc

```
@@ -80,6 +237,71 @@ class FaissIVFIndex::Adapter : public faiss::InvertedLists {
   void resize(size_t /* list_no */, size_t /* new_size */) override {
     assert(false);
   }
+
+ private:
+  class IteratorAdapter : public faiss::InvertedListsIterator {
+   public:
+    IteratorAdapter(KNNIterator* it, size_t list_no, size_t code_size)
+        : it_(it),
+          secondary_index_it_(it->GetSecondaryIndexIterator()),
+          code_size_(code_size) {
+      assert(it_);
+      assert(secondary_index_it_);
+
+      const std::string label = SerializeLabel(list_no);
+      secondary_index_it_->Seek(label);
+      Update();
+    }
+
+    bool is_available() const override { return id_and_codes_.has_value(); }
+
+    void next() override {
+      secondary_index_it_->Next();
+      Update();
+    }
+
+    std::pair<faiss::idx_t, const uint8_t*> get_id_and_codes() override {
+      assert(is_available());
+
+      return *id_and_codes_;
+    }
+
+   private:
+    void Update() {
+      id_and_codes_.reset();
+
+      const Status status = secondary_index_it_->status();
+      if (!status.ok()) {
+        throw std::runtime_error(status.ToString());
```

I wonder why we throw for errors here (and below). Wouldn't it be sufficient if `is_available()` returns false after `Update()` fails?

---

We have to differentiate between "out of items" and "some kind of error happened", similarly to RocksDB iterators. We wouldn't want to silently ignore I/O errors for example and produce an incorrect result.

---

For a bit more context, `IteratorAdapter`'s code executes in the context of FAISS, and FAISS uses exceptions to communicate errors. However, we eventually do catch all exceptions coming from FAISS and turn them into error statuses (regardless of whether the exception originated from FAISS itself or from our class).